docs/developers/local-rules.md

```diff
@@ -0,0 +1,196 @@
++│               ├── MyRule1.scala
++│               ├── MyRule2.scala
++│               └── ...
+ ├── service2
+ │   └── ...
+ └── service3
+     └── ...
+```
+
+```diff
+ // build.sbt
+ lazy val service1 = project
++  .settings(
++    libraryDependencies +=
++      "ch.epfl.scala" %%
++        "scalafix-core" %
++        _root_.scalafix.sbt.BuildInfo.scalafixVersion %
++        ScalafixConfig
++  )
+```
+
+### As a separate sub-project
+
+Declaring your rules in a separate sub-project allows you to use
+[scalafix-testkit](tutorial.md#write-unit-tests) to easily unit-test your
+semantic rules.
+
+```diff
+ repository
+ ├── build.sbt
+ ├── project
+ │   └── plugins.sbt
++├── scalafix // same source structure as the scalafix.g8 project template
++│   ├── input/src/main/scala/fix
++│   │   ├── Test1.scala
++│   │   ├── Test2.scala
++│   │   └── ...
++│   ├── output/src/main/scala/fix
++│   │   ├── Test1.scala
++│   │   ├── Test2.scala
++│   │   └── ...
++│   ├── rules/src
++│   │   └── main
++│   │       ├── resources/META-INF/services
++│   |       │   └── scalafix.v1.Rule // not mandatory, but allows syntax
++|   |       |                        // `MyRule1` over `class:fix.MyRule1`
++│   │       └── scala/fix
++│   │           ├── MyRule1.scala
++│   │           ├── MyRule2.scala
++│   │           └── ...
++│   └── tests/src/test/scala/fix
++│       └── RuleSuite.scala
+ ├── service1
+ │   └── ...
+ ├── service2
+ │   └── ...
+ └── service3
+     └── ...
+```
+
+```diff
+ // build.sbt
+ lazy val service1 = project // sub-project where rule(s) will be used
++  .dependsOn(`scalafix-rules` % ScalafixConfig)
++lazy val `scalafix-input` = (project in file("scalafix/input"))
++  .disablePlugins(ScalafixPlugin)
++lazy val `scalafix-output` = (project in file("scalafix/output"))
++  .disablePlugins(ScalafixPlugin)
++lazy val `scalafix-rules` = (project in file("scalafix/rules"))
++  .disablePlugins(ScalafixPlugin)
++  .settings(
++    libraryDependencies +=
++      "ch.epfl.scala" %%
++        "scalafix-core" %
++        _root_.scalafix.sbt.BuildInfo.scalafixVersion
++  )
++lazy val `scalafix-tests` = (project in file("scalafix/tests"))
++  .settings(
++    libraryDependencies +=
++      "ch.epfl.scala" %
++        "scalafix-testkit" %
++        _root_.scalafix.sbt.BuildInfo.scalafixVersion %
++        Test cross CrossVersion.full,
++    scalafixTestkitOutputSourceDirectories :=
++      sourceDirectories.in(`scalafix-output`, Compile).value,
++    scalafixTestkitInputSourceDirectories :=
++      sourceDirectories.in(`scalafix-input`, Compile).value,
++    scalafixTestkitInputClasspath :=
++      fullClasspath.in(`scalafix-input`, Compile).value,
++    scalafixTestkitInputScalacOptions :=
++      scalacOptions.in(`scalafix-input`, Compile).value,
++    scalafixTestkitInputScalaVersion :=
++      scalaVersion.in(`scalafix-input`, Compile).value
++  )
++  .dependsOn(`scalafix-input`, `scalafix-rules`)
++  .enablePlugins(ScalafixTestkitPlugin)
```

Comment:
diff is nice but hard to copy/paste, so I wonder if it's not better to leave it raw when 99% of the things are new, and the snippet is big

Reply:
I think we could have the final file here (instead of the diff). 
By the way, I like your comments/questions you add on your own PR, I find this quite useful and helpful for the review.

Reply:
> By the way, I like your comments/questions you add on your own PR, I find this quite useful and helpful for the review.

Thanks! However, sometimes they give context that should be captured in the git history as a source/commit comment instead, so I am trying to always question them since they can be an anti-pattern for long-term maintainability.